### PR TITLE
BI-9947 Fix issues with Kafka resilience in Item Handler

### PR DESCRIPTION
**Pull request by phudson-ch**
Improve item-handler resilience:

* Add retryable, non-retryable and ItemHandlerRuntimeException exceptions and hierarchy to exception package
* Modify OrderProcessorService to return response object
* Implement Visitor pattern to handle responses from OrderProcessorService
* Create an order service response handler and implement in order message consumers
* Implement handler logic where message is processed successfully; orders service is unavailable or orders service has returned an error
* Configure retry and error topics and maximum number of retries via system properties
* Refactor order message consumers into separate default, retry and error order message consumers
* Extract test synchronisation logic out from application controllers into corresponding test scoped Aspects (AOP)

Relates to: [BI-9947 Fix issues with Kafka resilience in Item Handler](https://companieshouse.atlassian.net/browse/BI-9947)

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/itemhandler/kafka/ItemMessageFactory.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
 
     private static final Logger LOGGER = LoggingUtils.getLogger();
 
-    private static final String CHD_ITEM_ORDERED_TOPIC = "chd-item-ordered";
+    public static final String CHD_ITEM_ORDERED_TOPIC = "chd-item-ordered";
```
Comment (cgregoryCH) [inline src/main/java/uk/gov/companieshouse/itemhandler/kafka/ItemMessageFactory.java:34]: should be private

File: src/test/java/uk/gov/companieshouse/itemhandler/kafka/ItemMessageProducerTest.java [hunk 2]
```diff
@@ -82,143 +65,123 @@
...
+//    private static void setFinalStaticField(Class<?> clazz, String fieldName, Object value)
+//            throws ReflectiveOperationException {
+//        final Field field = clazz.getDeclaredField(fieldName);
+//        field.setAccessible(true);
+//        final Field modifiers = Field.class.getDeclaredField("modifiers");
+//        modifiers.setAccessible(true);
+//        modifiers.setInt(field, field.getModifiers() & ~Modifier.FINAL);
+//        field.set(null, value);
+//    }
+
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/itemhandler/kafka/ItemMessageProducerTest.java:89]: remove commented out code

File: src/test/java/uk/gov/companieshouse/itemhandler/kafka/ItemMessageProducerTest.java [hunk 3]
```diff
@@ -82,143 +65,123 @@
...
     @Test
     @DisplayName("sendMessage delegates message creation to ItemMessageFactory")
     void sendMessageDelegatesMessageCreation() {
 
         // When
-        messageProducerUnderTest.sendMessage(ORDER, ORDER_REFERENCE, MISSING_IMAGE_DELIVERY_ITEM_ID);
+        messageProducerUnderTest.sendMessage(ORDER,
+                ORDER_REFERENCE,
+                MISSING_IMAGE_DELIVERY_ITEM_ID);
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/itemhandler/kafka/ItemMessageProducerTest.java:97]: formatting?

File: src/test/java/uk/gov/companieshouse/itemhandler/kafka/MessageDeserialiserTest.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+@ExtendWith(MockitoExtension.class)
+class MessageDeserialiserTest {
+    @Test
+    void deserializeReturnsNullIfMessageCannotBeDeserialised() {
+        assertNull(new MessageDeserialiser<>(email_send.class).deserialize("email-send",
+                "Test data".getBytes()));
+    }
+
+    //TODO: positive test for successful deserialisation
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/itemhandler/kafka/MessageDeserialiserTest.java:18]: TODO

File: src/test/java/uk/gov/companieshouse/itemhandler/kafka/OrderMessageConsumerTest.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+//        // Given & When
+//        when(serializerFactory.getGenericRecordSerializer(OrderReceived.class)).thenReturn(serializer);
+//        when(serializer.toBinary(any())).thenReturn(new byte[4]);
+//        ordersKafkaConsumer.republishMessageToTopic(ORDER_RECEIVED_URI,
+//                ORDER_RECEIVED_TOPIC_RETRY,
+//                ORDER_RECEIVED_TOPIC_ERROR);
+//        // Then
+//        verify(messageProducer, times(1)).sendMessage(any());
+//    }
+
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/itemhandler/kafka/OrderMessageConsumerTest.java:145]: remove comments

**Review by cgregoryCH — CHANGES_REQUESTED** [review]

**Review by cgregoryCH — APPROVED** [review]

**Review by kinpang-CH — APPROVED** [review]

**Review by dlgroves-ch — APPROVED** [review]

**Pull request merged**